### PR TITLE
Schema type table

### PR DESCRIPTION
**Pull request by andreamlin**
Using a DiscoTypeTable for storing and escaping symbols for DiscoGapic code gen.

Herein lies boilerplate for the DiscoGapic analogs to Model* (e.g. ModelTypeTable, ModelTypeNameConverter) classes.

### CODE REVIEW

**Comment by andreamlin** [comment]
PTAL

File: src/main/java/com/google/api/codegen/transformer/DiscoTypeTable.java [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+import com.google.api.codegen.discovery.Schema;
+import com.google.api.codegen.util.TypeAlias;
+import com.google.api.codegen.util.TypeTable;
+import java.util.Map;
+
+/**
+ * A DiscoTypeTable manages the imports for a set of fully-qualified type names, and provides helper
+ * methods for importing instances of TypeRef.
+ */
+public class DiscoTypeTable {
```
Comment (garrettjonesgoogle) [inline src/main/java/com/google/api/codegen/transformer/DiscoTypeTable.java]: Why not `SchemaTypeTable`?

File: src/test/java/com/google/api/codegen/testdata/discoveries/java/java_cloudbilling.v1.json.discogapic.baseline [hunk 2]
```diff
@@ -94,7 +94,7 @@ public final class BillingAccount implements Serializable {
   }
 
   @Override
-  public String toString() {
+  public java.lang.String toString() {
```
Comment (andreamlin) [inline src/test/java/com/google/api/codegen/testdata/discoveries/java/java_cloudbilling.v1.json.discogapic.baseline]: Done.

File: src/main/java/com/google/api/codegen/transformer/java/JavaSchemaTypeNameConverter.java [hunk 3]
```diff
@@ -0,0 +1,247 @@
...
+import com.google.api.codegen.util.TypeNameConverter;
+import com.google.api.codegen.util.TypedValue;
+import com.google.api.codegen.util.java.JavaTypeTable;
+import com.google.common.base.Strings;
+
+/** The Schema TypeName converter for Java. */
+public class JavaSchemaTypeNameConverter implements SchemaTypeNameConverter {
+
+  /** The package prefix protoc uses if no java package option was provided. */
+  private static final String DEFAULT_JAVA_PACKAGE_PREFIX = "com.google.protos";
```
Comment (andreamlin) [inline src/main/java/com/google/api/codegen/transformer/java/JavaSchemaTypeNameConverter.java]: Changing to "com.google.discovery", which is the package for the Google Discovery module. Any other name suggestions welcome.

File: src/main/java/com/google/api/codegen/transformer/java/JavaSchemaTypeNameConverter.java [hunk 4]
```diff
@@ -0,0 +1,247 @@
...
+
+  private static String getPrimitive(Schema schema) {
```
Comment (andreamlin) [inline src/main/java/com/google/api/codegen/transformer/java/JavaSchemaTypeNameConverter.java]: Done.

File: src/main/java/com/google/api/codegen/transformer/java/JavaSchemaTypeNameConverter.java [hunk 5]
```diff
@@ -0,0 +1,247 @@
...
+    }
+    if (primitiveType.equals("String")) {
+      return "\"\"";
+    }
+    throw new IllegalArgumentException("Schema is of unknown type.");
+  }
+
+  private TypeNameConverter typeNameConverter;
+
+  public JavaSchemaTypeNameConverter(String implicitPackageName) {
```
Comment (andreamlin) [inline src/main/java/com/google/api/codegen/transformer/java/JavaSchemaTypeNameConverter.java]: Done.

File: src/main/java/com/google/api/codegen/transformer/DiscoTypeTable.java (hunk 1 above)
Comment (andreamlin) [inline src/main/java/com/google/api/codegen/transformer/DiscoTypeTable.java]: Refactoring to SchemaTypeTable for now. I'm not sure right now if or how this class can be reused for tracking Discovery methods or resources too.

File: src/main/java/com/google/api/codegen/transformer/SchemaTypeNameConverter.java [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+ * limitations under the License.
+ */
+package com.google.api.codegen.transformer;
+
+import com.google.api.codegen.config.FieldConfig;
+import com.google.api.codegen.discovery.Schema;
+import com.google.api.codegen.util.TypeName;
+import com.google.api.codegen.util.TypedValue;
+
+/** SchemaTypeNameConverter maps TypeRef instances to TypeName instances. */
```
Comment (andreamlin) [inline src/main/java/com/google/api/codegen/transformer/SchemaTypeNameConverter.java]: Done.

File: src/main/java/com/google/api/codegen/transformer/java/JavaSchemaTypeNameConverter.java [hunk 7]
```diff
@@ -0,0 +1,247 @@
...
+
+  @Override
+  public TypeName getTypeNameForResourceNameElementType(
+      FieldConfig fieldConfig, String typedResourceShortName) {
+    // TODO(andrealin)
+    return null;
+  }
+
+  public static String getJavaPackage(Document file) {
+    String packageName = String.format("com.google.%s.%s", file.name(), file.version());
```
Comment (andreamlin) [inline src/main/java/com/google/api/codegen/transformer/java/JavaSchemaTypeNameConverter.java]: Putting this task down as a TODO for a later PR.

File: src/main/java/com/google/api/codegen/transformer/java/JavaSchemaTypeNameConverter.java [hunk 8]
```diff
@@ -0,0 +1,247 @@
...
+        // because those types are already boxed.
+        return typeNameConverter.getTypeName(primitiveTypeName);
+      } else {
+        if (shouldBoxPrimitives) {
+          return new TypeName(JavaTypeTable.getBoxedTypeName(primitiveTypeName));
+        } else {
+          return new TypeName(primitiveTypeName);
+        }
+      }
+    } else if (schema.type() == Type.ARRAY) {
```
Comment (andreamlin) [inline src/main/java/com/google/api/codegen/transformer/java/JavaSchemaTypeNameConverter.java]: Yes. This is a problem I'm addressing in a subsequent PR.

**Comment by andreamlin** [comment]
Addressed comments from yesterday; fixed up the minor changes, left TODOs for major changes.

**Comment by garrettjonesgoogle** [comment]
LGTM, @vchudnov-g do you want to take a look?

File: src/main/java/com/google/api/codegen/discogapic/transformer/java/JavaDiscoGapicNamer.java [hunk 9]
```diff
@@ -0,0 +1,25 @@
+/* Copyright 2016 Google Inc
```
Comment (vchudnov-g) [inline src/main/java/com/google/api/codegen/discogapic/transformer/java/JavaDiscoGapicNamer.java]: date

File: src/main/java/com/google/api/codegen/transformer/SchemaTypeFormatter.java [hunk 10]
```diff
@@ -0,0 +1,37 @@
+/* Copyright 2016 Google Inc
```
Comment (vchudnov-g) [inline src/main/java/com/google/api/codegen/transformer/SchemaTypeFormatter.java]: date

File: src/main/java/com/google/api/codegen/transformer/SchemaTypeFormatterImpl.java [hunk 11]
```diff
@@ -0,0 +1,46 @@
+/* Copyright 2016 Google Inc
```
Comment (vchudnov-g) [inline src/main/java/com/google/api/codegen/transformer/SchemaTypeFormatterImpl.java]: date

File: src/main/java/com/google/api/codegen/transformer/SchemaTypeNameConverter.java [hunk 12]
```diff
@@ -0,0 +1,63 @@
+/* Copyright 2016 Google Inc
```
Comment (vchudnov-g) [inline src/main/java/com/google/api/codegen/transformer/SchemaTypeNameConverter.java]: date

File: src/main/java/com/google/api/codegen/transformer/SchemaTypeTable.java [hunk 13]
```diff
@@ -0,0 +1,99 @@
+/* Copyright 2016 Google Inc
```
Comment (vchudnov-g) [inline src/main/java/com/google/api/codegen/transformer/SchemaTypeTable.java]: date

File: src/main/java/com/google/api/codegen/transformer/java/JavaSchemaTypeNameConverter.java [hunk 14]
```diff
@@ -0,0 +1,249 @@
+/* Copyright 2016 Google Inc
```
Comment (vchudnov-g) [inline src/main/java/com/google/api/codegen/transformer/java/JavaSchemaTypeNameConverter.java]: date

File: src/main/java/com/google/api/codegen/discogapic/DiscoGapicInterfaceContext.java [hunk 15]
```diff
@@ -16,49 +16,58 @@
...
 @AutoValue
 public abstract class DiscoGapicInterfaceContext implements InterfaceContext {
   public static DiscoGapicInterfaceContext create(
       Document document,
       GapicProductConfig productConfig,
-      ModelTypeTable typeTable,
-      SurfaceNamer namer,
+      SchemaTypeTable typeTable,
+      DiscoGapicNamer discoGapicNamer,
+      SurfaceNamer surfaceNamer,
```
Comment (andreamlin) [inline src/main/java/com/google/api/codegen/discogapic/DiscoGapicInterfaceContext.java]: The DiscoGapicNamer will simply render entities like the getter and setter method, whereas the existing SurfaceNamer class does more protobuf Model rendering. The SurfaceNamer is part of the interface that this class implements, so the SurfaceNamer isn't really used right now. Once gax is integrated, the SurfaceNamer will probably have methods to make all the Callable and Settings entities.

File: src/main/java/com/google/api/codegen/transformer/SchemaTypeTable.java (hunk 13 above)
Comment (andreamlin) [inline src/main/java/com/google/api/codegen/transformer/SchemaTypeTable.java]: Done.

File: src/main/java/com/google/api/codegen/transformer/java/JavaSchemaTypeNameConverter.java (hunk 14 above)
Comment (andreamlin) [inline src/main/java/com/google/api/codegen/transformer/java/JavaSchemaTypeNameConverter.java]: Done.

File: src/main/java/com/google/api/codegen/transformer/SchemaTypeNameConverter.java (hunk 12 above)
Comment (andreamlin) [inline src/main/java/com/google/api/codegen/transformer/SchemaTypeNameConverter.java]: Done.

File: src/main/java/com/google/api/codegen/transformer/SchemaTypeFormatterImpl.java (hunk 11 above)
Comment (andreamlin) [inline src/main/java/com/google/api/codegen/transformer/SchemaTypeFormatterImpl.java]: Done.

File: src/main/java/com/google/api/codegen/transformer/SchemaTypeFormatter.java (hunk 10 above)
Comment (andreamlin) [inline src/main/java/com/google/api/codegen/transformer/SchemaTypeFormatter.java]: Done.

File: src/main/java/com/google/api/codegen/discogapic/transformer/java/JavaDiscoGapicNamer.java (hunk 9 above)
Comment (andreamlin) [inline src/main/java/com/google/api/codegen/discogapic/transformer/java/JavaDiscoGapicNamer.java]: Done.

File: src/main/java/com/google/api/codegen/discogapic/transformer/DiscoGapicNamer.java [hunk 16]
```diff
@@ -0,0 +1,43 @@
+/* Copyright 2016 Google Inc
```
Comment (andreamlin) [inline src/main/java/com/google/api/codegen/discogapic/transformer/DiscoGapicNamer.java]: Done.

**Comment by andreamlin** [comment]
A schema is just a model, yep. "Model", capital 'M', is used in toolkit to either mean (1) a model in MVVM, or (2) a protobuf message class, afaik.

The rendered schema message class just has private fields, a builder class, and getters.

**Pull request merged**